File: server/src/security.js
import { db } from './db.js';
import { getSetting } from './settings.js';

// ---------- 客户端 IP ----------
export function getClientIp(req) {
  const xff = req.headers['x-forwarded-for'];
  if (typeof xff === 'string' && xff.length) return xff.split(',')[0].trim();
  if (Array.isArray(xff) && xff.length) return String(xff[0]).trim();
  return req.ip || req.socket?.remoteAddress || 'unknown';
}

// ---------- IP 白名单 ----------
function ipToLong(ip) {
  const m = /^(\d{1,3})\.(\d{1,3})\.(\d{1,3})\.(\d{1,3})$/.exec(ip);
  if (!m) return null;
  return ((+m[1] << 24) >>> 0) + (+m[2] << 16) + (+m[3] << 8) + +m[4];
}

function inCidr(ip, cidr) {
  const [net, bitsStr] = cidr.split('/');
  const bits = bitsStr ? parseInt(bitsStr, 10) : 32;
  const ipLong = ipToLong(ip);
  const netLong = ipToLong(net);
  if (ipLong === null || netLong === null || bits < 0 || bits > 32) return false;
  const mask = bits === 0 ? 0 : (0xffffffff << (32 - bits)) >>> 0;
  return (ipLong & mask) === (netLong & mask);
}

// 支持：精确 IP、CIDR（如 10.0.0.0/24）、前缀（如 192.168.1.）
export function ipInWhitelist(ip, listStr) {
  if (!listStr || !listStr.trim()) return true; // 空 = 不限制
  const list = listStr.split(',').map((s) => s.trim()).filter(Boolean);
  for (const item of list) {
    if (item === ip) return true;
    if (item.endsWith('.')) {
      if (ip.startsWith(item)) return true;
    } else if (item.includes('/')) {
      if (inCidr(ip, item)) return true;
    }
  }
  return false;
}

export function isIpAllowed(req) {
  const wl = getSetting('ip_whitelist');
  if (!wl || !wl.trim()) return true;
  return ipInWhitelist(getClientIp(req), wl);
}

// ---------- 登录失败锁定（防爆破，内存态） ----------
const fails = new Map(); // username -> { count, until }

export function getLockUntil(username) {
  const e = fails.get(username);
  if (!e) return 0;
  if (e.until > Date.now()) return e.until;
  // 仅清理真正过期的锁定（until>0 且已过期）；until=0 表示尚未锁定，仅用于计数，不能删，否则失败次数无法累加
  if (e.until > 0) fails.delete(username);
  return 0;
}

export function recordFail(username) {
  const max = getSetting('login_fail_max') || 5;
  const e = fails.get(username) || { count: 0, until: 0 };
  e.count += 1;
  if (e.count >= max) {
    const lockMin = getSetting('login_lock_minutes') || 15;
    e.until = Date.now() + lockMin * 60 * 1000;
  }
  fails.set(username, e);
  return e;
}

export function clearFail(username) {
  fails.delete(username);
}

// ---------- 审计日志 ----------
export function logAudit({ action, username = '', ip = '', success = true, detail = '' }) {
  try {
    db.prepare(
      `INSERT INTO audit_logs (action, username, ip, success, detail, created_at)
       VALUES (?, ?, ?, ?, ?, datetime('now'))`
    ).run(action, username, ip, success ? 1 : 0, detail);
  } catch {
    /* 审计失败不应影响主流程 */
  }
}

export function recentAudit(limit = 100) {
  const n = Math.min(Math.max(parseInt(limit, 10) || 100, 1), 500);
  return db.prepare('SELECT * FROM audit_logs ORDER BY id DESC LIMIT ?').all(n);
}

File: server-go/internal/security/security.go
package security

import (
	"net"
	"net/http"
	"strconv"
	"strings"
	"sync"
	"time"

	"terminal-web/server-go/internal/db"
	"terminal-web/server-go/internal/settings"
)

// ---------- 客户端 IP ----------

func GetClientIP(r *http.Request) string {
	if xff := r.Header.Get("X-Forwarded-For"); xff != "" {
		parts := strings.Split(xff, ",")
		return strings.TrimSpace(parts[0])
	}
	host, _, err := net.SplitHostPort(r.RemoteAddr)
	if err != nil {
		return r.RemoteAddr
	}
	return host
}

// ---------- IP 白名单 ----------

func ipToLong(ip string) (uint32, bool) {
	parts := strings.Split(ip, ".")
	if len(parts) != 4 {
		return 0, false
	}
	var l uint32
	for i := 0; i < 4; i++ {
		n, err := strconv.Atoi(parts[i])
		if err != nil || n < 0 || n > 255 {
			return 0, false
		}
		l = (l << 8) + uint32(n)
	}
	return l, true
}

func inCidr(ip, cidr string) bool {
	idx := strings.Index(cidr, "/")
	if idx < 0 {
		return false
	}
	netPart := cidr[:idx]
	bitsStr := cidr[idx+1:]
	bits, err := strconv.Atoi(bitsStr)
	if err != nil || bits < 0 || bits > 32 {
		return false
	}
	ipLong, ok1 := ipToLong(ip)
	netLong, ok2 := ipToLong(netPart)
	if !ok1 || !ok2 {
		return false
	}
	var mask uint32
	if bits == 0 {
		mask = 0
	} else {
		mask = (0xffffffff << (32 - bits)) & 0xffffffff
	}
	return (ipLong & mask) == (netLong & mask)
}

// IpInWhitelist 支持精确 IP、CIDR、前缀（以 "." 结尾）。
func IpInWhitelist(ip, listStr string) bool {
	if strings.TrimSpace(listStr) == "" {
		return true
	}
	for _, item := range strings.Split(listStr, ",") {
		item = strings.TrimSpace(item)
		if item == "" {
			continue
		}
		if item == ip {
			return true
		}
		if strings.HasSuffix(item, ".") {
			if strings.HasPrefix(ip, item) {
				return true
			}
		} else if strings.Contains(item, "/") {
			if inCidr(ip, item) {
				return true
			}
		}
	}
	return false
}

func IsIPAllowed(r *http.Request) bool {
	wl := settings.GetString("ip_whitelist")
	if strings.TrimSpace(wl) == "" {
		return true
	}
	return IpInWhitelist(GetClientIP(r), wl)
}

// ---------- 登录失败锁定（内存态） ----------

type lockEntry struct {
	count int
	until int64
}

var (
	failsMu sync.Mutex
	fails   = map[string]*lockEntry{}
)

// GetLockUntil 返回锁定解除的时间戳（毫秒）；未锁定返回 0。
func GetLockUntil(username string) int64 {
	failsMu.Lock()
	defer failsMu.Unlock()
	e, ok := fails[username]
	if !ok {
		return 0
	}
	now := time.Now().UnixMilli()
	if e.until > now {
		return e.until
	}
	if e.until > 0 {
		delete(fails, username)
	}
	return 0
}

func RecordFail(username string) {
	failsMu.Lock()
	defer failsMu.Unlock()
	max := settings.GetInt("login_fail_max")
	if max <= 0 {
		max = 5
	}
	e := fails[username]
	if e == nil {
		e = &lockEntry{}
		fails[username] = e
	}
	e.count++
	if e.count >= max {
		lockMin := settings.GetInt("login_lock_minutes")
		if lockMin <= 0 {
			lockMin = 15
		}
		e.until = time.Now().UnixMilli() + int64(lockMin)*60*1000
	}
}

func ClearFail(username string) {
	failsMu.Lock()
	defer failsMu.Unlock()
	delete(fails, username)
}

// ---------- 审计日志 ----------

func LogAudit(action, username, ip string, success bool, detail string) {
	_, err := db.DB.Exec(
		`INSERT INTO audit_logs (action, username, ip, success, detail, created_at)
		 VALUES (?, ?, ?, ?, ?, datetime('now'))`,
		action, username, ip, boolToInt(success), detail)
	if err != nil {
		// 审计失败不影响主流程
		return
	}
}

func boolToInt(b bool) int {
	if b {
		return 1
	}
	return 0
}

func RecentAudit(limit int) ([]map[string]interface{}, error) {
	if limit < 1 {
		limit = 100
	}
	if limit > 500 {
		limit = 500
	}
	rows, err := db.DB.Query("SELECT * FROM audit_logs ORDER BY id DESC LIMIT ?", limit)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	cols, err := rows.Columns()
	if err != nil {
		return nil, err
	}
	out := []map[string]interface{}{}
	for rows.Next() {
		vals := make([]interface{}, len(cols))
		ptrs := make([]interface{}, len(cols))
		for i := range vals {
			ptrs[i] = &vals[i]
		}
		if err := rows.Scan(ptrs...); err != nil {
			return nil, err
		}
		m := map[string]interface{}{}
		for i, c := range cols {
			if b, ok := vals[i].([]byte); ok {
				m[c] = string(b)
			} else {
				m[c] = vals[i]
			}
		}
		out = append(out, m)
	}
	return out, nil
}
